### PR TITLE
E1959.Intelligent copying of assignments without topics (Issue #1341)

### PR DESCRIPTION
**Pull request by subhasekhar**
we identified the code to make the changes and identified the pattern used.

### CODE REVIEW

File: app/views/assignments/checktopicscopy.html.erb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<h1>Select the desired choice and copy the assignment</h1>
+<%= form_tag(:controller => :assignments, :action => :copy, :id => @assignment_id) do%>
+  <%= radio_button_tag(:copyoption, "copyWithoutTopics") %>
+  <%= label_tag(:copyoption_topics, "copy without topics") %><br>
+  <%= radio_button_tag(:copyoption, "copyWithTopics") %>
+  <%= label_tag(:copyoption_topics, "copy with topics") %><br>
+  <%= radio_button_tag(:copyoption, "copyWithTopicsTeams") %>
+  <%= label_tag(:copyoption_topics_teams, "copy with topics and teams") %><br>
+  <%= submit_tag("Create") %>
```
Comment (Yashad) [inline app/views/assignments/checktopicscopy.html.erb]: Shouldn't the submit button name be "Copy"

File: app/assets/javascripts/tree_display.jsx [hunk 2]
```diff
@@ -633,561 +640,562 @@ jQuery(document).ready(function() {
...
+    })
+
+    if (document.getElementById("tree_display")) {
+        React.render(
+            React.createElement(TabSystem),
+            document.getElementById("tree_display")
+        )
+    }
 
 })
```
Comment (Yashad) [inline app/assets/javascripts/tree_display.jsx]: Seems like you did not intend to change this file but introduced many whitespaces and newlines unknowingly. Please correct it. (remove this file from commit if it does not have any changes relevant to your project  or correct its code climate )

File: config/routes.rb [hunk 3]
```diff
@@ -143,7 +144,7 @@
     end
   end
 
-resources :institution, except: [:destroy] do
+  resources :institution, except: [:destroy] do
```
Comment (Yashad) [inline config/routes.rb]: white space

File: spec/controllers/assignments_controller_spec.rb [hunk 4]
```diff
@@ -306,11 +306,11 @@
         expect(response).to redirect_to('/assignments/2/edit')
       end
     end
-
-    context 'when new assignment id does not fetch successfully' do
-      it 'shows an error flash message and redirects to assignments#edit page' do
-        allow(assignment).to receive(:dup).and_return(new_assignment)
-        allow(new_assignment).to receive(:save).and_return(false)
+	
+	context 'when new assignment is not able to be copied' do
```
Comment (Yashad) [inline spec/controllers/assignments_controller_spec.rb]: white spaces

**Comment by Yashad** [comment]
Do you have any other pull request for expertiza beta? We would like these changes on beta.

File: app/models/assignment_form.rb [hunk 5]
```diff
@@ -340,14 +342,15 @@ def self.copy(assignment_id, user)
...
+          user_id: user.id,
+          notification_limit: aq.notification_limit,
+          questionnaire_weight: aq.questionnaire_weight,
+          used_in_round: aq.used_in_round,
+          dropdown: aq.dropdown
       )
     end
   end
 end
+
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply app/models/assignment_form.rb]: Removed.

File: app/models/assignment_form.rb [hunk 6]
```diff
@@ -339,15 +350,15 @@ def self.copy(assignment_id, user)
...
-        assignment_id: new_assign.id,
-        questionnaire_id: aq.questionnaire_id,
-        user_id: user.id,
-        notification_limit: aq.notification_limit,
-        questionnaire_weight: aq.questionnaire_weight,
-        used_in_round: aq.used_in_round,
-        dropdown: aq.dropdown
-      )
+        AssignmentQuestionnaire.create(
+            assignment_id: new_assign.id,
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply app/models/assignment_form.rb]: Solved.

File: app/models/assignment_form.rb [hunk 7]
```diff
@@ -339,15 +350,15 @@ def self.copy(assignment_id, user)
...
-      AssignmentQuestionnaire.create(
-        assignment_id: new_assign.id,
-        questionnaire_id: aq.questionnaire_id,
-        user_id: user.id,
-        notification_limit: aq.notification_limit,
-        questionnaire_weight: aq.questionnaire_weight,
-        used_in_round: aq.used_in_round,
-        dropdown: aq.dropdown
-      )
+        AssignmentQuestionnaire.create(
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply app/models/assignment_form.rb]: Solved.

File: spec/models/assignment_form_spec.rb [hunk 8]
```diff
@@ -310,13 +310,26 @@
...
+      expect(AssignmentForm.copy(1, 'copyWithoutTopics', build(:instructor))).to eq(2)
     end
+
+    it 'copies the original assignment to a new one and returns the new assignment_id when copyoption equals copyWithTopicTeams' do
+      allow(Assignment).to receive(:find).with(1).and_return(assignment)
+      allow_any_instance_of(AssignmentForm).to receive(:copy_assignment_questionnaire).with(any_args).and_return('OK!')
+      allow(AssignmentDueDate).to receive(:copy).with(1, any_args).and_return('OK!')
+      allow_any_instance_of(Assignment).to receive(:create_node).and_return('OK!')
+      allow(SignUpTopic).to receive(:where).with(assignment_id: 1).and_return([build(:topic)])
+	  allow_any_instance_of(Assignment).to receive(:create_node).and_return('OK!')
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb [hunk 9]
```diff
@@ -310,13 +310,26 @@
...
+	  allow_any_instance_of(AssignmentForm).to receive(:update_attribute).with(any_args).and_return('OK!')
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb [hunk 10]
```diff
@@ -310,13 +310,26 @@
...
+      expect(AssignmentForm.copy(1, 'copyWithTopicsTeams', build(:instructor))).to eq(2)
+    end
+
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb [hunk 11]
```diff
@@ -310,13 +310,26 @@
...
+      allow_any_instance_of(AssignmentForm).to receive(:copy_assignment_questionnaire).with(any_args).and_return('OK!')
+      allow(AssignmentDueDate).to receive(:copy).with(1, any_args).and_return('OK!')
+      allow_any_instance_of(Assignment).to receive(:create_node).and_return('OK!')
+      allow(SignUpTopic).to receive(:where).with(assignment_id: 1).and_return([build(:topic)])
+	  allow_any_instance_of(Assignment).to receive(:create_node).and_return('OK!')
+	  allow_any_instance_of(AssignmentForm).to receive(:update_attribute).with(any_args).and_return('OK!')
+      expect(AssignmentForm.copy(1, 'copyWithTopicsTeams', build(:instructor))).to eq(2)
+    end
+	
+
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/models/assignment_form_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/models/assignment_form_spec.rb]: Solved.

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/controllers/assignments_controller_spec.rb]: Solved.

File: spec/controllers/assignments_controller_spec.rb [hunk 12]
```diff
@@ -306,11 +306,11 @@
...
+      it 'should show an error and redirect to assignments#tree display page' do	  
+		allow(AssignmentForm).to receive(:copy).with("1", nil, instructor).and_return(false)
+        allow(Assignment).to receive(:find).with(2).and_return(false)
```
[the remark this answers is not in the document]
Comment (rohanpillai20) [reply spec/controllers/assignments_controller_spec.rb]: Solved.

File: spec/controllers/assignments_controller_spec.rb (hunk 4 above)
Comment (rohanpillai20) [reply spec/controllers/assignments_controller_spec.rb]: Solved.

**Pull request merged**